### PR TITLE
Complete several TODOs

### PR DESCRIPTION
**Pull request by EgorBu**
* handle custom modification functions as well as regexp 
https://github.com/src-d/style-analyzer/issues/526#issuecomment-454801024
* use dulwich in the future
https://github.com/src-d/style-analyzer/issues/526#issuecomment-454804489
* switch to numpy arrays
https://github.com/src-d/style-analyzer/issues/526#issuecomment-454808654
https://github.com/src-d/style-analyzer/issues/526#issuecomment-454809317

### CODE REVIEW

File: lookout/style/format/benchmarks/generate_smoke.py [hunk 1]
```diff
@@ -103,14 +104,20 @@ def generate_smoke_entry(inputpath: str, outputpath: str, force: bool = False) -
         writer = csv.DictWriter(index_file, fieldnames=["repo", "style", "from", "to"])
         writer.writeheader()
         for style_name, (pattern, repl) in js_format_rules.items():
-            # TODO(zurk): handle custom modification functions as well as regexp.
-            pattern = re.compile(pattern)
+            if isinstance(pattern, str):
```
Comment (zurk) [inline lookout/style/format/benchmarks/generate_smoke.py]: I am not sure about its signature (input parameters mostly). For now, I think that `repl ` argument is redundant and it should look like 
```
new_style_code = modify_style(initial_code)
```

I think we can keep this TODO in code as soon as we do not have any function for style modification.
When (and If) we actually need style modification functions, we can resolve it with a better understanding of what we actually need.

If we want to close all todos in the code we can just delete it. I think it is a minor one.

File: lookout/style/format/benchmarks/top_repos_quality.py [hunk 2]
```diff
@@ -136,10 +137,8 @@ def ensure_repo(repository: str, storage_dir: str) -> str:
     if os.path.exists(repository):
         return repository
     # clone repository
-    # TODO: use dulwich in the future
     git_dir = os.path.join(storage_dir, get_repo_name(repository))  # location for code
-    cmd = "git clone --single-branch --branch master %s %s" % (repository, git_dir)
-    subprocess.check_call(cmd.split())
+    porcelain.clone(repository, git_dir)
```
Comment (zurk) [inline lookout/style/format/benchmarks/top_repos_quality.py]: it is not possible to clone only `--single-branch` via Dulwich, isn't it?

File: lookout/style/format/benchmarks/top_repos_quality.py [hunk 3]
```diff
@@ -211,26 +210,18 @@ def wrapped_capture_quality_report(*args, **kwargs):
 
 def calc_weighted_avg(arr: Sequence[Sequence], col: int, weight_col: int = 5) -> float:
     """Calculate average value in `col` weighted by column `weight_col`."""
-    numerator, denominator = 0, 0
-    # TODO: switch to numpy arrays
-    for row in arr:
-        numerator += row[col] * row[weight_col]
-        denominator += row[weight_col]
+    arr = numpy.array(arr)
```
Comment (zurk) [inline lookout/style/format/benchmarks/top_repos_quality.py]: `arr` here can contain ints, floats as well as strings. Does this code work well in this case?

File: lookout/style/format/benchmarks/generate_smoke.py [hunk 4]
```diff
@@ -103,14 +104,20 @@ def generate_smoke_entry(inputpath: str, outputpath: str, force: bool = False) -
...
+                pattern_ = re.compile(pattern)
+
+                def pattern(repl, code):
+                    return re.sub(pattern_, repl, code)
+            else:
+                assert isinstance(pattern, FunctionType), "Expected function, got %s" % \
```
[the remark this answers is not in the document]
Comment (EgorBu) [reply lookout/style/format/benchmarks/generate_smoke.py]: https://github.com/src-d/style-analyzer/pull/535#discussion_r248968605

File: lookout/style/format/benchmarks/top_repos_quality.py (hunk 3 above)
Comment (EgorBu) [reply lookout/style/format/benchmarks/top_repos_quality.py]: `.astype(float)` added to be sure that it works.

Comment (zurk) [reply lookout/style/format/benchmarks/top_repos_quality.py]: ok, I hope it is fine now.

**Pull request merged**